mysql/access_rules.go

```diff
@@ -0,0 +1,294 @@
+	Operation         AccessRuleOperation `json:"operation"`
+	Status            AccessRuleStatus    `json:"status"`
+	PollInterval      time.Duration       `json:"-"`
+	Timeout           time.Duration       `json:"-"`
+}
+
+func (c *AccessRulesClient) CreateAccessRule(input *CreateAccessRuleInput) error {
+
+	if input.ServiceInstanceID != "" {
+		c.ServiceInstanceID = input.ServiceInstanceID
+	}
+
+	if err := c.createAccessRuleResource(input, nil); err != nil {
+		return err
+	}
+
+	pollInterval := input.PollInterval
+	if pollInterval == 0 {
+		pollInterval = WaitForAccessRulePollInterval
+	}
+
+	timeout := input.Timeout
+	if timeout == 0 {
+		timeout = WaitForAccessRuleTimeout
+	}
+
+	getRuleInput := &GetAccessRuleInput{
+		ServiceInstanceID: input.ServiceInstanceID,
+		Name:              input.RuleName,
+	}
+
+	return c.WaitForAccessRuleReady(getRuleInput, pollInterval, timeout)
+}
+
+func (c *AccessRulesClient) GetAllAccessRules(input *GetAccessRuleInput) (*AccessRuleList, error) {
+
+	if input.ServiceInstanceID != "" {
+		c.ServiceInstanceID = input.ServiceInstanceID
+	}
+
+	var accessRules AccessRuleList
+	if err := c.getAccessRulesResource(&accessRules); err != nil {
+		return nil, err
+	}
+
+	// Iterated through entire slice, rule was not found.
+	// No error occurred though, return a nil struct, and allow the Provdier to handle
+	// a Nil response case.
+	return &accessRules, nil
+}
+
+func (c *AccessRulesClient) GetAccessRule(input *GetAccessRuleInput) (*AccessRuleInfo, error) {
+
+	if input.ServiceInstanceID != "" {
+		c.ServiceInstanceID = input.ServiceInstanceID
+	}
+
+	var accessRules AccessRuleList
+	if err := c.getAccessRulesResource(&accessRules); err != nil {
+		return nil, err
+	}
+
+	// This is likely not the most optimal path for this, however, the upper bound on
+	// performance here is the actual API request, not the iteration.
+	for _, rule := range accessRules.AccessRules {
+		if rule.RuleName == input.Name {
+			return &rule, nil
+		}
+	}
+
+	// Iterated through entire slice, rule was not found.
+	// No error occurred though, return a nil struct, and allow the Provdier to handle
+	// a Nil response case.
+	return nil, nil
+}
+
+func (c *AccessRulesClient) WaitForAccessRuleReady(input *GetAccessRuleInput, pollInterval time.Duration, timeoutSeconds time.Duration) error {
+
+	err := c.client.WaitFor("access rule to be created.", pollInterval, timeoutSeconds, func() (bool, error) {
+
+		var info AccessRuleList
+		if err := c.getAccessRulesResource(&info); err != nil {
+			return false, err
+		}
+
+		c.client.DebugLogString(fmt.Sprintf("[DEBUG] Checking Activities : %v", info))
+		for _, accessRule := range info.AccessRules {
+			if accessRule.RuleName == input.Name {
+				return true, nil
+			}
+		}
+
+		for _, activity := range info.Activities {
+			if activity.AccessRuleActivityInfo.RuleName == input.Name {
+
+				switch s := strings.ToUpper(activity.AccessRuleActivityInfo.Status); s {
```

Comment:
Let's add logs under each case so we know what the status is whenever we poll. This is especially useful under the default case when we don't know the specific status is.